README.md
sse-channel (Server-Sent Events channel)
========================================

[![Version npm](http://img.shields.io/npm/v/sse-channel.svg?style=flat-square)](http://browsenpm.org/package/sse-channel)[![Build Status](http://img.shields.io/travis/rexxars/sse-channel/main.svg?style=flat-square)](https://travis-ci.org/rexxars/sse-channel)[![Coverage Status](http://img.shields.io/coveralls/rexxars/sse-channel/main.svg?style=flat-square)](https://coveralls.io/r/rexxars/sse-channel?branch=main)[![Code Climate](http://img.shields.io/codeclimate/github/rexxars/sse-channel.svg?style=flat-square)](https://codeclimate.com/github/rexxars/sse-channel/)

SSE-implementation which can be used to any node.js http request/response stream.

# Features

  - Easily attach to any node.js http request ([express](examples/express/server.js), [hapi](examples/hapi/server.js), [node http](examples/node.js/server.js))
  - History is maintained automatically, max size is configurable
  - Optionally pre-populate history when creating the channel
  - Automatically sends missed events to clients when reconnecting
  - Attempts to keep clients alive by sending "pings" automatically
  - Easily send messages to all clients or to specific clients
  - Configurable reconnection timeout
  - Auto-encode packets as JSON (configurable)
  - Supports a [number](https://github.com/amvtek/EventSource) of [different](https://github.com/Yaffle/EventSource/) [polyfills](https://github.com/remy/polyfills/blob/master/EventSource.js)
  - Works with the [compression](https://github.com/expressjs/compression) middleware
  - If polyfilled on the client side, works down to IE8 and Android 2.x
  - Maintains active connection count per channel

# Installing

```
npm install --save sse-channel
```

# Basic usage

```js
var SseChannel = require('sse-channel');
var http = require('http');

// Set up an interval that broadcasts server date every second
var dateChannel = new SseChannel();
setInterval(function broadcastDate() {
    dateChannel.send((new Date()).toISOString());
}, 1000);

// Create a regular HTTP server (works with express, too)
http.createServer(function(req, res) {
    // Note that you can add any client to an SSE channel, regardless of path.
    // Only requirement is not having written data to the response stream yet
    if (req.url.indexOf('/channel/date') === 0) {
        dateChannel.addClient(req, res);
    } else {
        res.writeHead(404);
        res.end();
    }
}).listen(7788, '127.0.0.1', function() {
    console.log('Listening on http://127.0.0.1:7788/');
});
```

See [examples](examples/) for express/hapi/node server examples.

# Options

The following are available as options to the `SseChannel` constructor.

- `historySize` - The maximum number of messages to keep in history. Note: Only messages with an `id` property are saved to history. Default: `500`
- `history` - Array of messages to pre-populate the history with. Note: If the size of the array specified exceeds the specified `historySize`, the array will be sliced to contain only the last `X` elements. Default: `[]`
- `retryTimeout` - How many milliseconds clients should wait before attempting to reconnect, if disconnected. Default: `undefined` - browser default.
- `pingInterval` - How often the server should send a "ping" to clients in order to keep the connection alive (milliseconds). Default: `20000` (20s)
- `jsonEncode` - Whether the client should auto-encode messages as JSON before sending. Default: `false`.

# Methods

All instances of `SseChannel` have the following methods:

### channel.addClient(request, response, [callback])

Add a client to the channel. `request` is the `http.IncomingMessage` instance for the connection, while `response` is the `http.ServerResponse` instance to write data to.

If specified, the `callback` supplied will be called once the client has been added.

### channel.removeClient(response)

Removes a client (`response`) from a channel. Note that this does not actually close the connection - for that, use `response.end()`. Also worth mentioning is that most clients will see this as an unintentional disconnect and will thus attempt to reconnect. Developers are encouraged to implement some sort of `disconnect` event that can be sent prior to disconnecting clients.

### channel.getConnectionCount()

Get the number of active connections currently connected to the channel.

### channel.ping()

Manually broadcast a "ping" to all connected clients, to keep the connections alive. Note that this is handled automatically by specifying the `pingInterval` option when instantiating the channel.

### channel.retry(duration)

Tell clients how many milliseconds (`duration`) they should wait before attempting to reconnect, if disconnected.

### channel.send(msg, [clients])

Send a message to all (or a subset of) connected clients.

`msg` can either be a string or an object. Specifying just a string will send the message without any `id` or `event` name, and will not end up in the history for the channel. The alternative is to provide an object:
 - `msg.data` - Data to send to the client. Data will be JSON-encoded if `jsonEncode` is enabled for the channel, otherwise it is sent as-is.
 - `msg.id` - ID of the event. Used by clients when reconnecting to ensure all messages are received. You are required to specify this if you want the message to be added to the channel history and have clients request missed data on reconnects.
 - `msg.event` - Event name, used on the client side to trigger on specific events.
 - `msg.retry` - Retry timeout (same as `retry()`)

### channel.sendEventsSinceId(response, sinceId)

Send events since a given ID from the channel history to the specified client. Usually handled automatically on reconnect, but could be useful for more customized implementations.

### channel.close()

Close all connections on this channel and stop sending "keep alive"-pings. Note that this will usually only trigger all clients to reconnect, which is probably not what you want. You might want to implement some sort of `disconnect`-contract between the server and client, before you call this.

# Events

`SseChannel` implements the EventEmitter contract, and emits the following events:

* `connect` - When a client has successfully connected and the initial headers have been queued for writing. Listeners are provided with the following arguments: `channel`, `request`, `response`
* `disconnect` - When a client closes the connection for any reason. Listeners are provided with the following arguments: `channel`, `response`
* `message` - When a message has been queued for sending. Listeners are provided with the following arguments: `channel`, `msg`, `clients`. In this case, `clients` is an array of either the clients provided or the full array of clients.

# Client usage

In browsers, use `EventSource` as you normally would. For browsers that don't natively support `EventSource`, I recommend amvtek's [polyfill](https://github.com/amvtek/EventSource). Basic client usage:

```js
var es = new EventSource('http://localhost:1337/channels/soccer-goals');
es.addEventListener('goal', function(e) {
    alert('GOAL! ' + e.data);
});

// On the server:
soccerGoalChannel.send({
    event: 'goal',
    data: 'Manchester Utd - Swansea (2-1: Wayne Rooney)'
})
```

Note that the above is the usage when using named events. The default (and most basic) way is:

```js
var es = new EventSource('http://localhost:1337/channels/ticker');
es.onmessage = function(e) {
    alert(e.data);
};

// On the server:
tickerChannel.send({ data: 'Some message here' });
```

Also note that when using `jsonEncode` option of SSE-Channel, you'll have to decode the message on the client:

```js
var es = new EventSource('http://localhost:1337/channels/sysinfo');
es.onmessage = function(e) {
    var data = JSON.parse(e.data);
    console.log('CPU usage: ' + e.cpuUsage);
};

// On the server:
var sysInfoChannel = new SseChannel({ jsonEncode: true });
sysInfoChannel.send({
    data: { cpuUsage: 13.37 }
});
```

# Advanced example

```js
var SseChannel = require('sse-channel');
var http = require('http');
var os = require('os');

// Set up a new channel. Most of these options have sane defaults,
// feel free to look at lib/sse-channel.js for all available options
var sysInfoChannel = new SseChannel({
    retryTimeout: 250,
    historySize: 300,
    pingInterval: 15000,
    jsonEncode: true,
});

// Set up an HTTP server
http.createServer(function(req, res) {
    // The '/channel' prefix is not necessary - you can add any client
    // to an SSE-channel, if you want to.
    if (req.url.indexOf('/channel/sysinfo') === 0) {
        // Serve the client using sysinfo SSE-channel
        sysInfoChannel.addClient(req, res);
    } else {
        res.writeHead(404);
        res.end();
    }
}).listen(7788, '127.0.0.1', function() {
    console.log('Listening on http://127.0.0.1:7788/');
});

// Set up an interval that broadcasts system info every 250ms
var sysInfoCounter = 0;
setInterval(function broadcastSysInfo() {
    sysInfoChannel.send({
        id: sysInfoCounter++,
        data: {
            freemem: os.freemem(),
            loadavg: os.loadavg()
        }
    });
}, 250);
```

License
-------
MIT-licensed, see `LICENSE`


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [4.0.0] - 2020-11-19
- BREAKING: Only support Node v10 and higher
- BREAKING: Remove CORS-handling - use `cors` module or similar

## [3.1.0] - 2018-06-06
### Added
- Add ability to disable CORS-checking by passing `false` as the `cors` option

## [3.0.0] - 2017-11-27
- BREAKING: Only support Node v4 and higher

## [2.0.6] - 2017-03-29
### Fixed
- Stop "keep-alive" ping when calling `close()` on a channel

## [2.0.5] - 2017-02-10
### Fixed
- Fixed `flush()` deprecation warnings

## 2.0.4 - 2017-02-10
### Changed
- Content-Type header now includes charset (UTF-8)

## 2.0.3 - 2016-12-11
### Fixed
- Make connection count more reliable (William Neely)

## [2.0.0] - 2015-07-23
### Changed
- `jsonEncode` will now encode any message (previously, specifying a string to `send` would send a non-JSON encoded string)
- `addClient()` callback is now asyncronous in all cases, and provides an `Error` instance on CORS-failure
- `sendMissedEvents()` is renamed to `sendEventsSinceId()`
- `message` event now provides `channel` as first argument to listeners, to be consistent with `connect` and `disconnect` events

### Added
- A much richer README with proper documentation for options, methods and events
- Examples of usage with Node.js HTTP server, express.js and hapi

## [1.0.6] - 2015-07-23
### Fixed
- Call `flush()` on response object if it exists, fixes compression middleware in express

## [1.0.4] - 2015-02-03
### Fixed
- Use `0` instead of `Infinity` for socket timeouts.

### Added
- Changelog!

[4.0.0]: https://github.com/rexxars/sse-channel/compare/v3.1.0...v4.0.0
[3.1.0]: https://github.com/rexxars/sse-channel/compare/v3.0.0...v3.1.0
[3.0.0]: https://github.com/rexxars/sse-channel/compare/v2.0.6...v3.0.0
[2.0.6]: https://github.com/rexxars/sse-channel/compare/v2.0.5...v2.0.6
[2.0.5]: https://github.com/rexxars/sse-channel/compare/v2.0.4...v2.0.5
[2.0.0]: https://github.com/rexxars/sse-channel/compare/1.0.6...2.0.0
[1.0.6]: https://github.com/rexxars/sse-channel/compare/1.0.4...1.0.6
[1.0.4]: https://github.com/rexxars/sse-channel/compare/1.0.3...1.0.4


lib/sse-channel.js
'use strict';

var url = require('url');
var util = require('util');
var events = require('events');

// See initializeConnection() for an explanation
var preambleData = new Array(2057).join('-') + '\n';

/**
 * Server-Sent Events "Channel"
 *
 * The "channel"-concept is simply a collection of clients which should receive the same messages.
 * This is useful if you want to have one server which can easily handle different sources and
 * clients. History is automatically handled for you, as long as each event is accompanied by
 * a numeric ID. If a client reconnects with a `Last-Event-ID`-header, he receives all messages
 * newer than the given ID.
 *
 * @param {Object}  options               Options for this SSE-channel
 * @param {Array}   options.history       An array of messages to pre-populate the history with.
 *                                        Note: Number of items will equal the max history size,
 *                                        where the last elements in the array will be the present
 * @param {Number}  options.historySize   The maximum number of messages to keep in history
 * @param {Number}  options.retryTimeout  Milliseconds clients should wait before reconnecting
 * @param {Number}  options.pingInterval  How often the server should send a "ping" to clients
 * @param {Boolean} options.jsonEncode    Whether the client should auto-encode data as JSON before
 *                                        sending. Defaults to false.
 */
var SseChannel = function(options) {
    var opts = options || {};

    var jsonEncode = (
        typeof opts.jsonEncode === 'undefined' ?
            false :
            Boolean(opts.jsonEncode)
    );

    this.jsonEncode = jsonEncode;
    this.historySize = opts.historySize || 500;
    this.retryTimeout = opts.retryTimeout || null;
    this.pingInterval = (opts.pingInterval | 0) || 20000;

    // Populate history with the entries specified
    this.history = ((opts.history || [])
        .filter(hasId)
        .slice(0 - this.historySize)
        .reverse()
        .map(function(msg) {
            return {
                id: msg.id,
                msg: parseMessage(msg, jsonEncode)
            };
        })
    );

    this.connections = [];
    this.connectionCount = 0;

    // Start a timer that will ping all connected clients at a given interval
    this.timer = setInterval(this.ping.bind(this), this.pingInterval);
};

util.inherits(SseChannel, events.EventEmitter);

/**
 * Add a new client to the channel
 *
 * @param {Request}  req        Request of the client
 * @param {Response} res        Response of the client
 * @param {Function} [callback] Optional callback to run when the client has been added
 * @return {void}
 */
SseChannel.prototype.addClient = function(req, res, callback) {
    // amvtek's EventSource polyfill uses the query string to work around some limitations in
    // Internet Explorer. The last received event ID for instance will be a query parameter
    // instead of a header, and there's also a bug where IE will need 2kb of data before it
    // starts dispatching progress events: https://github.com/amvtek/EventSource/wiki/UserGuide
    var query = url.parse(req.url, true).query || {};

    // Initialize the connection
    initializeConnection({
        request: req,
        response: res,
        retry: this.retryTimeout,
        preamble: query.evs_preamble
    });

    // Add the connection to our pool
    this.connections.push(res);
    this.connectionCount = this.connections.length;

    // When the client disconnects, remove the client
    var removeClient = this.removeClient.bind(this, res);
    req.on('end', removeClient);
    req.on('close', removeClient);
    res.on('finish', removeClient);

    // The "last event id" is normally sent as a header,
    // but various polyfills apply it to the query string
    var lastEventId = (
        req.headers['last-event-id'] ||
        query.evs_last_event_id ||
        query.lastEventId ||
        0
    );

    // See if the client has requested some history entries
    if (lastEventId) {
        this.sendEventsSinceId(res, lastEventId);
    }

    this.emit('connect', this, req, res);

    if (callback) {
        process.nextTick(callback);
    }
};

/**
 * Remove the client from the channel
 *
 * @param {Response} res Response of the client
 */
SseChannel.prototype.removeClient = function(res) {
    var i = this.connections.indexOf(res)
    while (i !== -1) {
        this.connections.splice(i, 1)
        i = this.connections.indexOf(res)
    }

    this.connectionCount = this.connections.length;

    this.emit('disconnect', this, res);
};

/**
 * Get number of active connections on this channel
 *
 * @return {Number} Number of active connections
 */
SseChannel.prototype.getConnectionCount = function() {
    return this.connections.length;
};

/**
 * Send a "ping" (empty comment) to all clients, to keep the connections alive
 *
 */
SseChannel.prototype.ping = function() {
    broadcast(this.connections, ':\n');
};

/**
 * Tell the clients how long they should wait before reconnecting.
 * Note: This sends a message to all clients.
 *
 * @param {Number} retryTimeout Milliseconds clients should wait before reconnecting,
 *                              if they are disconnected.
 */
SseChannel.prototype.retry = function(retryTimeout) {
    this.retryTimeout = retryTimeout;
    broadcast(this.connections, 'retry: ' + retryTimeout + '\n');
};

/**
 * Send a message to all clients on the channel
 *
 * @param {Object|String} msg Message to send to the client
 * @param {String} msg.data   Data to send to the client (will be JSON-encoded if `jsonEncode` is enabled)
 * @param {Number} msg.id     ID of the event (used by clients when reconnecting to ensure all messages are received)
 * @param {String} msg.event  Event name (used on client side to trigger on specific event names)
 * @param {String} msg.retry  Retry timeout (same as `retry()`)
 * @param {Array}  [clients]  Optional array of Response objects - if specified, the message will
 *                            be sent only to these clients, as well as bypassing the history.
 */
SseChannel.prototype.send = function(msg, clients) {
    var message = parseMessage(msg, this.jsonEncode);

    if (!clients) {
        // Remove duplicate entries from the history
        if (msg.id) {
            this.history = this.history.filter(entry => entry.id !== msg.id)
        }

        // Add the message to history (if not a "private" message)
        var entry = { id: msg.id, msg: message };
        if (msg.id && this.history.unshift(entry) > this.historySize) {
            this.history.pop();
        }
    }

    broadcast(clients || this.connections, message);

    this.emit('message', this, msg, clients || this.connections);
};

/**
 * Send events since a given ID to the specified client
 *
 * @param  {Response} response Response object to use
 * @param  {Number}   sinceId  The event ID to use as a lower bound (non-inclusive)
 */
SseChannel.prototype.sendEventsSinceId = function(response, sinceId) {
    var entries = [];
    for (var i = 0; i < this.history.length; i++) {
        if (this.history[i].id <= sinceId) {
            break;
        }

        // Unshift so we send the oldest messages first
        entries.unshift(this.history[i].msg);
    }

    entries.map(function(msg) {
        response.write(msg);
    });

    flush(response);
};

/**
 * Close all connections on this channel and stop the ping timer
 *
 */
SseChannel.prototype.close = function() {
    clearInterval(this.timer);

    var i = this.connections.length;
    while (i--) {
        this.connections[i].end();
    }
};

/**
 * Sends the initial, required headers for the connection
 *
 * @param {Object}   opts           Options object
 * @param {Request}  opts.request   Request object to use
 * @param {Response} opts.response  Response object to use
 * @param {Number}   opts.retry     Time in milliseconds to specify as reconnection timeout
 * @param {Boolean}  opts.preamble  Whether to send a "preamble" of dummy data to the client
 */
function initializeConnection(opts) {
    opts.request.socket.setTimeout(0);
    opts.request.socket.setNoDelay(true);
    opts.request.socket.setKeepAlive(true);
    opts.response.writeHead(200, {
        'Content-Type': 'text/event-stream;charset=UTF-8',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive'
    });

    opts.response.write(':ok\n\n');

    // Only set retry if it has a sane value
    var retry = opts.retry | 0;
    if (retry) {
        opts.response.write('retry: ' + retry + '\n');
    }

    // Should we delivar a "preamble" to the client? Required in some cases by Internet Explorer,
    // see https://github.com/amvtek/EventSource/wiki/UserGuide for more information
    if (opts.preamble) {
        opts.response.write(':' + preambleData);
    }

    flush(opts.response);
}

/**
 * Broadcast a packet to all connected clients
 *
 * @param  {Array}  connections Array of connections (response instances) to write to
 * @param  {String} packet      The chunk of data to broadcast
 */
function broadcast(connections, packet) {
    var i = connections.length;
    while (i--) {
        connections[i].write(packet);
        flush(connections[i]);
    }
}

/**
 * Calls `flush()` on the response if it exists. This is necessary in the case
 * when you are using the `compression` middleware for express.js, where the
 * middleware will wait for a certain amount of data before sending it to the
 * client in order to compress in a more efficient manner.
 *
 * Unfortunately, node has a native (but deprecated) flush() method which will trigger
 * deprecation warnings if called. Thus, we're checking the name of the function.
 * Please open a pull request if you can come up with a cleaner approach to this.
 *
 * @param {Response} response Response object to flush
 */
function flush(response) {
    if (response.flush && response.flush.name !== 'deprecated') {
        response.flush();
    }
}

/**
 * Parse a message object (or string) into a writable data chunk
 *
 * @param  {String|Object} msg        Object or string to parse into sendable message
 * @param  {Boolean}       jsonEncode Whether to JSON-encode data parameter
 * @return {String}
 */
function parseMessage(msg, jsonEncode) {
    if (typeof msg === 'string') {
        msg = { data: msg };
    }

    var output = '';
    if (msg.event) {
        output += 'event: ' + msg.event + '\n';
    }

    if (msg.retry) {
        output += 'retry: ' + msg.retry + '\n';
    }

    if (msg.id) {
        output += 'id: ' + msg.id + '\n';
    }

    var data = msg.data || '';
    if (jsonEncode) {
        data = JSON.stringify(data);
    }

    output += parseTextData(data);

    return output;
}

/**
 * Parse text data, ensuring it doesn't break the SSE-protocol
 *
 * @param  {String} text
 * @return {String}
 */
function parseTextData(text) {
    var data = String(text).replace(/(\r\n|\r|\n)/g, '\n');
    var lines = data.split(/\n/), line;

    var output = '';
    for (var i = 0, l = lines.length; i < l; ++i) {
        line = lines[i];

        output += 'data: ' + line;
        output += (i + 1) === l ? '\n\n' : '\n';
    }

    return output;
}

/**
 * Returns whether a given item has an `id` property or not
 *
 * @param  {Object}  item
 * @return {Boolean}
 */
function hasId(item) {
    return item.id && item.id > 0;
}

module.exports = SseChannel;
